libbeat/common/fmtstr/formatevents.go

```diff
@@ -0,0 +1,276 @@
+		}
+	}
+	return fields
+}
+
+// Run executes the format string returning a new expanded string or an error
+// if execution or event field expansion fails.
+func (fs *EventFormatString) Run(event common.MapStr) (string, error) {
+	if err := fs.collectFields(event); err != nil {
+		return "", err
+	}
+	return fs.formatter.Run()
+}
+
+// Eval executes the format string, writing the resulting string into the provided output buffer. Returns error if execution or event field expansion fails.
+func (fs *EventFormatString) Eval(out *bytes.Buffer, event common.MapStr) error {
+	if err := fs.collectFields(event); err != nil {
+		return err
+	}
+	return fs.formatter.Eval(out)
+}
+
+// collectFields tries to extract and convert all required fields into an array
+// of strings.
+func (fs *EventFormatString) collectFields(event common.MapStr) error {
+	for i, fi := range fs.fields {
+		s, err := fieldString(event, fi.path)
+		if err != nil {
+			if fi.required {
+				return err
+			}
+
+			s = ""
+		}
+		fs.ctx.keys[i] = s
+	}
+
+	return nil
+}
+
+func (e *eventFieldCompiler) compileEventField(
+	field string,
+	ops []VariableOp,
+) (FormatEvaler, error) {
+	if len(ops) > 1 {
+		return nil, errors.New("Too many format modifiers given")
+	}
+
+	defaultValue := ""
+	if len(ops) == 1 {
+		op := ops[0]
+		if op.op != ":" {
+			return nil, fmt.Errorf("unsupported format operator: %v", op.op)
+		}
+		defaultValue = op.param
+	}
+
+	path, err := parseEventPath(field)
+	if err != nil {
+		return nil, err
+	}
+
+	info, found := e.keys[path]
+	if !found {
+		info = keyInfo{
+			required: len(ops) == 0,
+			index:    e.index,
+		}
+		e.index++
+		e.keys[path] = info
+	} else if !info.required && len(ops) == 0 {
+		info.required = true
+		e.keys[path] = info
+	}
+
+	idx := info.index
+
+	if len(ops) == 0 {
+		return &eventFieldEvaler{e.ctx, idx}, nil
+	}
+
+	return &defaultEventFieldEvaler{e.ctx, idx, defaultValue}, nil
+}
+
+func (e *eventFieldEvaler) Eval(out *bytes.Buffer) error {
+	type stringer interface {
+		String() string
+	}
+
+	s := e.ctx.keys[e.index]
+	_, err := out.WriteString(s)
+	return err
+}
+
+func (e *defaultEventFieldEvaler) Eval(out *bytes.Buffer) error {
+	type stringer interface {
+		String() string
+	}
+
+	s := e.ctx.keys[e.index]
+	if s == "" {
+		s = e.defaultValue
+	}
+	_, err := out.WriteString(s)
+	return err
+}
+
+func parseEventPath(field string) (string, error) {
+	field = strings.Trim(field, " \n\r\t")
+	var fields []string
+
+	for len(field) > 0 {
+		if field[0] != '[' {
+			return "", errors.New("expected field extractor start with '['")
+		}
+
+		idx := strings.IndexByte(field, ']')
+		if idx < 0 {
+			return "", errors.New("missing closing ']'")
+		}
+
+		path := field[1:idx]
+		if path == "" {
+			return "", errors.New("empty fields selector '[]'")
+		}
+
+		fields = append(fields, path)
+		field = field[idx+1:]
+	}
+
+	path := strings.Join(fields, ".")
+	return path, nil
+}
+
+// TODO: move to libbeat/common?
```

Comment:
not done. `common.MapStr` does not directly support querying a field as string-value. Same method might become use-full when extracting compound keys from events (e.g. think lookup-cache)